File: packages/shared/src/schemas/target-app.ts
import { z } from "zod";
import { UUIDSchema, DateTimeSchema, PaginationParamsSchema, PaginatedResponseSchema } from "./common";

export const TargetAppTypeSchema = z.enum(["juice_shop", "dvwa", "custom"]);
export type TargetAppType = z.infer<typeof TargetAppTypeSchema>;

export const TargetAppStatusSchema = z.enum(["pending", "ready", "error", "deprecated"]);
export type TargetAppStatus = z.infer<typeof TargetAppStatusSchema>;

export const TargetAppConfigSchema = z.object({
  baseUrl: z.string().url().optional(),
  dockerImage: z.string().optional(),
  dockerTag: z.string().default("latest"),
  environment: z.record(z.string()).default({}),
  healthCheckPath: z.string().default("/"),
  healthCheckInterval: z.number().int().positive().default(30),
  resetScript: z.string().optional(),
  exposedPorts: z.array(z.number().int().positive()).default([80, 443]),
  resources: z.object({
    cpuLimit: z.string().optional(),
    memoryLimit: z.string().optional()
  }).optional()
});
export type TargetAppConfig = z.infer<typeof TargetAppConfigSchema>;

export const TargetAppSchema = z.object({
  id: UUIDSchema,
  projectId: UUIDSchema,
  name: z.string().min(1).max(255),
  type: TargetAppTypeSchema,
  config: TargetAppConfigSchema,
  status: TargetAppStatusSchema.default("pending"),
  lastValidatedAt: DateTimeSchema.optional(),
  validationError: z.string().optional(),
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema
});
export type TargetApp = z.infer<typeof TargetAppSchema>;

export const CreateTargetAppRequestSchema = z.object({
  projectId: UUIDSchema,
  name: z.string().min(1).max(255),
  type: TargetAppTypeSchema,
  config: TargetAppConfigSchema
});
export type CreateTargetAppRequest = z.infer<typeof CreateTargetAppRequestSchema>;

export const UpdateTargetAppRequestSchema = z.object({
  name: z.string().min(1).max(255).optional(),
  config: TargetAppConfigSchema.partial().optional(),
  status: TargetAppStatusSchema.optional()
}).refine(data => Object.keys(data).length > 0, {
  message: "At least one field must be provided"
});
export type UpdateTargetAppRequest = z.infer<typeof UpdateTargetAppRequestSchema>;

export const ValidateTargetAppRequestSchema = z.object({
  targetAppId: UUIDSchema
});
export type ValidateTargetAppRequest = z.infer<typeof ValidateTargetAppRequestSchema>;

export const ValidateTargetAppResponseSchema = z.object({
  success: z.boolean(),
  status: TargetAppStatusSchema,
  error: z.string().optional(),
  details: z.record(z.unknown()).optional()
});
export type ValidateTargetAppResponse = z.infer<typeof ValidateTargetAppResponseSchema>;

export const TargetAppListParamsSchema = PaginationParamsSchema.extend({
  projectId: UUIDSchema.optional(),
  type: TargetAppTypeSchema.optional(),
  status: TargetAppStatusSchema.optional()
});
export type TargetAppListParams = z.infer<typeof TargetAppListParamsSchema>;

export const TargetAppListResponseSchema = PaginatedResponseSchema(TargetAppSchema);
export type TargetAppListResponse = z.infer<typeof TargetAppListResponseSchema>;

File: apps/api/app/schemas/target_app.py
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field, HttpUrl

from app.schemas.common import PaginationParams


class TargetAppConfig(BaseModel):
    base_url: Optional[HttpUrl] = None
    docker_image: Optional[str] = None
    docker_tag: str = "latest"
    environment: dict = Field(default_factory=dict)
    health_check_path: str = "/"
    health_check_interval: int = Field(default=30, ge=1)
    reset_script: Optional[str] = None
    exposed_ports: list[int] = Field(default_factory=lambda: [80, 443])
    resources: Optional[dict] = None


class TargetAppBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    type: str = Field(pattern="^(juice_shop|dvwa|custom)$")
    config: TargetAppConfig


class TargetAppCreate(TargetAppBase):
    project_id: UUID


class TargetAppUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    config: Optional[TargetAppConfig] = None
    status: Optional[str] = Field(default=None, pattern="^(pending|ready|error|deprecated)$")


class TargetAppResponse(TargetAppBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    status: str
    last_validated_at: Optional[datetime] = None
    validation_error: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class TargetAppListParams(PaginationParams):
    project_id: Optional[UUID] = None
    type: Optional[str] = Field(default=None, pattern="^(juice_shop|dvwa|custom)$")
    status: Optional[str] = Field(default=None, pattern="^(pending|ready|error|deprecated)$")


class ValidateTargetAppRequest(BaseModel):
    target_app_id: UUID


class ValidateTargetAppResponse(BaseModel):
    success: bool
    status: str
    error: Optional[str] = None
    details: Optional[dict] = None


class TargetAppListResponse(BaseModel):
    items: list[TargetAppResponse]
    total: int
    page: int
    limit: int
    total_pages: int